lib/core/src/Cardano/Wallet/Primitive/Slotting.hs

```diff
@@ -55,123 +66,116 @@ import Cardano.Wallet.Primitive.Types
     ( neverForksSummary )
 
 import qualified Cardano.Slotting.Slot as Cardano
 import qualified Ouroboros.Consensus.BlockchainTime.WallClock.Types as Cardano
 import qualified Ouroboros.Consensus.HardFork.History.Qry as HF
 
--- -----------------------------------------------------------------------------
--- New Api using ouroboros-consensus. With the right interpreter, the
--- calculations don't break on hard-forks.
-
-
---
--- Queries
---
+{-------------------------------------------------------------------------------
+                                    Queries
+-------------------------------------------------------------------------------}
 
-currentEpoch :: MonadIO m => TimeInterpreter m -> m (Maybe EpochNo)
-currentEpoch ti = ti . epochAt =<< liftIO getCurrentTime
+-- | This is 'Ouroboros.Consensus.HardFork.History.Qry.Qry' wrapped in a reader
+-- to provide the blockchain system start time as context.
+type Qry = ReaderT StartTime HF.Qry
 
-epochAt :: UTCTime -> Qry (Maybe EpochNo)
-epochAt = traverse epochOf <=< ongoingSlotAt
+-- | Query the blockchain start time. This is part of the 'TimeInterpreter'
+-- environment.
+getStartTime :: Qry StartTime
+getStartTime = ask
 
+-- | Query the epoch corresponding to a flat slot number.
 epochOf :: SlotNo -> Qry EpochNo
 epochOf slot = epochNumber <$> toSlotId slot
 
+-- | Query to convert a flat 'SlotNo' to a 'SlotId', which is the epoch number,
+-- and the local slot index.
 toSlotId :: SlotNo -> Qry SlotId
-toSlotId slot = HardForkQry $ do
-    (e, s, _) <- HF.slotToEpoch slot
+toSlotId slot = do
+    (e, s) <- lift $ slotToEpoch' slot
     return $ SlotId
         (EpochNo $ fromIntegral $ Cardano.unEpochNo e)
-        (SlotInEpoch $ unsafeConvert s)
+        (SlotInEpoch $ downCast s)
   where
-    unsafeConvert :: Word64 -> Word32
-    unsafeConvert = fromIntegral
+    downCast :: Word64 -> Word32
+    downCast = fromIntegral
 
-startTime :: SlotNo -> Qry UTCTime
-startTime s = do
-    rel <- HardForkQry (fst <$> HF.slotToWallclock s)
-    RelToUTCTime rel
+-- | Query the absolute time at which a slot starts.
+slotToUTCTime :: SlotNo -> Qry UTCTime
+slotToUTCTime sl = slotToRelTime sl >>= fromRelativeTime
 
--- | Can be used to know when the next epoch starts.
+-- | Query the relative time at which a slot starts.
+slotToRelTime :: SlotNo -> Qry RelativeTime
+slotToRelTime = lift . fmap fst . slotToWallclock
+
+-- | Query the absolute times at which an epoch starts and ends.
 --
--- This is preferable to asking for the start time of the /next/ epoch, because
--- the next epoch may be outside the forecast range, and result in
--- @PastHorizonException@.
-endTimeOfEpoch :: EpochNo -> Qry UTCTime
-endTimeOfEpoch epoch = do
+-- Querying the end time of /this/ epoch is preferable to querying the start
+-- time of the /next/ epoch, because the next epoch may be outside the forecast
+-- range, and result in 'PastHorizonException'.
+timeOfEpoch :: EpochNo -> Qry (UTCTime, UTCTime)
```

Comment:
So `timeOfEpoch` returns the _end_ time, and `slotToUTCTime` returns the _start_ time? I think this is confusing.

Shouldn't this function be keep it this function as `endTimeOfEpoch`?

Reply:
Ah, sorry nvm — just noticed the `(UTCTime, UTCTime)`